packages/frontend/src/Happy/Frontend.hs

```diff
@@ -0,0 +1,57 @@
+module Happy.Frontend where
+  
+import Happy.Grammar
+import Happy.Frontend.AbsSyn
+import Happy.Frontend.Mangler
+import Happy.Frontend.PrettyGrammar
+import Happy.Frontend.Parser
+import Happy.Frontend.ParseMonad.Class
+import System.Exit
+import System.IO
+
+parseYFileContents :: String -> ParseResult AbsSyn
+parseYFileContents contents = runFromStartP ourParser contents 1
+
+mangleAbsSyn :: AbsSyn -> String -> ParseResult Grammar
+mangleAbsSyn abssyn filename = first unlines (mangler filename abssyn)
+  where first f (Left a) = Left (f a)
+        first _ (Right b) = Right b
+
+writePrettyFile :: Maybe String -> AbsSyn -> IO ()
+writePrettyFile location abssyn = do
+  let out = render (ppAbsSyn abssyn) in
+    case location of
+      Just s   -> writeFile s out >>
+        hPutStrLn stderr ("Production rules written to: " ++ s)
+      Nothing  -> return ()
+
+possDelitify :: String -> String -> IO (String, String)
+possDelitify ('y':'l':'.':nm) fl = return (deLitify fl, reverse nm)
+possDelitify ('y':'.':nm) fl     = return (fl, reverse nm)
+possDelitify f            _      = die ("`" ++ reverse f ++ "' does not end in `.y' or `.ly'\n")
+#if !MIN_VERSION_base(4,8,0)
+  where die s = hPutStr stderr s >> exitWith (ExitFailure 1) 
+#endif
```

Comment:
This function isn't ready to come out of the exe yet. It cuts across concerns to weirdly.

If you want, you can make a function here to parse the filename (including doing the reversing), and it should return `Maybe Bool` or something instead of any `IO`. Then the exe can use that, do the `die`-ing, and de lit the file according to the `Bool`.